### ISSUE TITLE
How to convert Tensor back to BitMap or any image format in Android?

### ISSUE THREAD

**Issue by nauyan** [pytorch/pytorch#30655]
I have converted a PyTorch model for Android mobile. The purpose of the model is to achieve Super Resolution. The problem I am facing is that the model gives output in the form of Tensor. Whereas I want to convert that tensor into some imaging format but I haven't been able to find a method to achieve this task. 

I cannot find something suitable in Pytorch Java documentation for this certain task. Please advise regarding this issue.

**Comment by ljk53** [comment]
You can call Tensor.getRawDataBuffer() API to get the underlying data buffer of output tensor: https://pytorch.org/docs/master/org/pytorch/Tensor.html#getrawdatabuffer
Or one of these API to obtain Java array (but this way it will allocate the buffer on Java heap and copy data, which will be slower): https://pytorch.org/docs/master/org/pytorch/Tensor.html#getdataasfloatarray
 
Once you have access to the raw pixels you can use other Java libraries or SDK API to convert the pixels into images.

cc: @IvanKobzarev

**Comment by phillies** [comment]
Here's a gist that converts a tensor to a RGBA bitmap with grayscale values:
https://gist.github.com/phillies/830f52b0bf592c32fc507669694f66d8
You can get a floatArray from a tensor using `myTensor.dataAsFloatArray`

**Comment by AbdelsalamHaa** [comment]
Hi I have the exact same issue. @ljk53 I have tried ur suggested answers but the Tensor.getRawDataBuffer() function is a private function therefore I can't use it. 
I can call getDataAsFloatArray() but I have no idea How to convert a float to a bitmap. 
When I try to call getDataAsBufferArray() the program will crash because my Tesnor is a float.

**Comment by AbdelsalamHaa** [comment]
Hi, I have tried the suggest solution but it didn't work for me. Tensor.getRawDataBuffer() is a private function therefore I can't really use it. 
I can call Tensor.getDataAsFloatArray() but I don't know how to convert to a bitmap from there. 
I also tried to call Tensor.getDataAsBufferArray() but my tensor is a type float so it will cause my program to crash.

**Comment by AbdelsalamHaa** [comment]
Hi Guys, after so long of trying I manged to do it. I have converted the Tensor to a float than I converted this code to java and it worked. 
the code is below 
Note: this is to convert to RGB Bitmap 

Here is the output tensor:


    final Tensor outputTensor = module.forward(IValue.from(inputTensor)).toTensor();
    float[] array1 = outputTensor.getDataAsFloatArray() ;

    for (int i = 0 ; i< array1.length ; i++){
      arralist.add(array1[i]);
    }
    Bitmap outBitmap = arrayFlotToBitmap(arralist,420,276);

// here is the function 

    private Bitmap arrayFlotToBitmap(List<Float>  floatArray,int width,int height){

    byte alpha = (byte) 255 ;

    Bitmap bmp = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888) ;

    ByteBuffer byteBuffer = ByteBuffer.allocate(width*height*4*3) ;

    float Maximum = Collections.max(floatArray);
    float minmum = Collections.min(floatArray);
    float delta = Maximum - minmum ;

    int i = 0 ;
    for (float value : floatArray){
      byte temValue = (byte) ((byte) ((((value-minmum)/delta)*255)));
      byteBuffer.put(4*i, temValue) ;
      byteBuffer.put(4*i+1, temValue) ;
      byteBuffer.put(4*i+2, temValue) ;
      byteBuffer.put(4*i+3, alpha) ;
      i++ ;
    }
    bmp.copyPixelsFromBuffer(byteBuffer) ;
    return bmp ;
  }

**Comment by Achilleasein** [comment]
Are you sure about the arralist.add() part? shouldnt it specify the position?
like: arralist.add( i , array1[i] );

**Comment by GeghamHayrapetian** [comment]
please help me . I want to use this code but it throw this exception java.lang.IndexOutOfBoundsException: index=640000 out of bounds (limit=640000).

**Comment by Achilleasein** [comment]
> please help me . I want to use this code but it throw this exception java.lang.IndexOutOfBoundsException: index=640000 out of bounds (limit=640000).

Would you like me to send you the codes I have created?

**Comment by sepideh-srj** [comment]
I used this code in my application but when I show in the image component, it's just a line of black and white pixels. Any idea what I'm doing wrong? 
I'd really appreciate any help on this.

**Comment by sepideh-srj** [comment]
> I have converted a PyTorch model for Android mobile. The purpose of the model is to achieve Super Resolution. The problem I am facing is that the model gives output in the form of Tensor. Whereas I want to convert that tensor into some imaging format but I haven't been able to find a method to achieve this task.
> 
> I cannot find something suitable in Pytorch Java documentation for this certain task. Please advise regarding this issue.



> > please help me . I want to use this code but it throw this exception java.lang.IndexOutOfBoundsException: index=640000 out of bounds (limit=640000).
> 
> Would you like me to send you the codes I have created?

Hey is it possible that I also see your code?

**Comment by nauyan** [comment]
@sepideh-srj haven't been able to make any progress since this issue so had to put this work on hold.

**Comment by Achilleasein** [comment]
Instead of posting my code I will give you the repos that have been used, since my code is not well written and those repos do the job with proper commenting: 
https://github.com/pytorch/android-demo-app
https://github.com/onnx/tutorials/blob/master/tutorials/PytorchCaffe2MobileSqueezeNet.ipynb
https://github.com/tensorflow/examples/tree/master/lite/examples/image_classification/android
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/android
https://www.tensorflow.org/lite/examples
https://github.com/googlecodelabs/tensorflow-for-poets-2

If you still have issues I'll post my code.

**Comment by sepideh-srj** [comment]
> Instead of posting my code I will give you the repos that have been used, since my code is not well written and those repos do the job with proper commenting:
> https://github.com/pytorch/android-demo-app
> https://github.com/onnx/tutorials/blob/master/tutorials/PytorchCaffe2MobileSqueezeNet.ipynb
> https://github.com/tensorflow/examples/tree/master/lite/examples/image_classification/android
> https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/android
> https://www.tensorflow.org/lite/examples
> https://github.com/googlecodelabs/tensorflow-for-poets-2
> 
> If you still have issues I'll post my code.

Thank you,
So did you use pytorch for your work or did you convert it to tensorflow?
because the repos you sent are for converting to caffe and tensorflow and I didnt find anything on the pytorch repo.
Could you please post your code?

**Comment by curehabit** [comment]
> please help me . I want to use this code but it throw this exception java.lang.IndexOutOfBoundsException: index=640000 out of bounds (limit=640000).

I meet the exception too. Have you resolve it?

**Comment by Silence-Monk** [comment]
// I modified the code from **phillies** to get coloful image. Note that the format of an output tensor is typically NCHW.
// Here's my function in Kotlin. Hopefully it works in your case:

```
private fun floatArrayToBitmap(floatArray: FloatArray, width: Int, height: Int) : Bitmap {

        // Create empty bitmap in ARGB format
        val bmp: Bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val pixels = IntArray(width * height * 4)

        // mapping smallest value to 0 and largest value to 255
        val maxValue = floatArray.max() ?: 1.0f
        val minValue = floatArray.min() ?: -1.0f
        val delta = maxValue-minValue

        // Define if float min..max will be mapped to 0..255 or 255..0
        val conversion = { v: Float -> ((v-minValue)/delta*255.0f).roundToInt()}

        // copy each value from float array to RGB channels and set alpha channel
        for (i in 0 until width * height) {
            val r = conversion(floatArray[i])
            val g = conversion(floatArray[i+width*height])
            val b = conversion(floatArray[i+2*width*height])
            pixels[i] = rgb(r, g, b) // you might need to import for rgb()
        }
        bmp.setPixels(pixels, 0, width, 0, 0, width, height)

        return bmp
    }
```

**Comment by roj4s** [comment]
I'm using this code for colorfull images (based on previous answers) ...

public static Bitmap bitmapFromRGBImageAsFloatArray(float[] data, int width, int height){

        Bitmap bmp = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        float max = 0;
        float min = 999999;

        for(float f: data){
            if(f > max){
                max = f;
            }
            if(f < min){
                min = f;
            }
        }

        int delta = (int)(max - min);

        for (int i = 0; i < width * height; i++) {


            int r = (int) ((data[i] - min)/delta*255.0f);
            int g = (int) ((data[i + width * height] - min) / delta*255.0f);
            int b = (int) ((data[i + width * height * 2] - min) / delta*255.0f);

            int x = i / width;
            int y = i % width;

            int color = Color.rgb(r, g, b);
            bmp.setPixel(x, y, color);

        }
        return bmp;
    }

**Comment by m-bortnikov** [comment]
I want to share my version of the tensor to bitmap conversion. It is based on [AbdelsalamHaa](https://github.com/AbdelsalamHaa) code from this issue, which as I understand, intended to visulaize grayscale masks, e.g., after image segmentation. I changed only the order in which output pixels are set, because in the original version all three (R, G and B) channels were set to the same pixel values, returning basically a monochrome image.

```private Bitmap arrayFlotToBitmap(List<Float>  floatArray,int width,int height){

byte alpha = (byte) 255 ;

Bitmap bmp = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888) ;

ByteBuffer byteBuffer = ByteBuffer.allocate(width*height*4*3) ;

float Maximum = Collections.max(floatArray);
float minmum = Collections.min(floatArray);
float delta = Maximum - minmum ;

int i = 0 ;
for (float value : floatArray){
  byte temValue = (byte) ((byte) ((((value-minmum)/delta)*255)));
  if (i < width * height) {byteBuffer.put(4 * i, temValue);}  // red channel
  else {if (i < 2 * width * height) {byteBuffer.put(4 * (i - width * height) + 1, temValue);}  // green channel
  else {byteBuffer.put(4 * (i - 2 * width * height) + 2, temValue);}}  // blue channel
  byteBuffer.put(4*i+3, alpha) ;
  i++ ;
}
bmp.copyPixelsFromBuffer(byteBuffer) ;
return bmp ;
}```
